### ISSUE TITLE
Search transaction usability

### ISSUE THREAD

**Issue by feri42** [cosmos/gaia#75]
**Issue:**
Using the `GET /txs` gaia-lite API call:

1) It is currently not possible to search for all transactions for a given address.
2) It is not possible to resume a search at some given point without fetching transactions that have already been fetched.
3) It is not possible to only fetch the latest transactions.

**Description:**
1) A common usage pattern for transaction searches is querying for all transactions for a given address. As it stands now we must make two requests to the server once for the recipient once for the sender. This effectively doubles the request load on the server. It also makes it more complicated on the client side where the two sets of transactions must be merged.
2) A common usage pattern is also to have a cursor to continue the search where we left off last time. For example if we fetch transactions up to block N, then next time we would like to start fetching at block N + 1. As it stands now we must use the existing paging implementation where we use the page as a cursor. The downside is that we must fetch the last page over and over again.
3) Yet another pattern is to only fetch a few most recent transactions. As it stands now we must always start with the oldest transactions because there is no way to guess on what page the most recent transactions can be found.

**Proposed solution**
1) Add an optional `address` query parameter that searches for recipient OR sender.
2) Add optional start block and/or start time/end time parameters
3) Add a search direction parameter, or failing that reverse the return order to be from latest to oldest, which is much more useful.

**Comment by alexanderbez** [comment]
> 1) It is currently not possible to search for all transactions for a given address.

Search for all txs that were sent by some sender? If so, this is possible. `/txs?mesage.sender={addr}`.

> 2) It is not possible to resume a search at some given point without fetching transactions that have already been fetched.

Pagination should allow you to do this.

> 3) It is not possible to only fetch the latest transactions.

Define latest transactions?

**Comment by feri42** [comment]
> Search for all txs that were sent by some sender?
No, search for messages sent by some address OR recevied to the same address. Think of an address as an account.

> Pagination should allow you to do this.
Only in a very sloppy fashion. Say I finished on page 5 at the last request, now I need to re-fetch page 5 every tome to see if something was added.

> Define latest transactions?
Think "most recent N transactions". Even most recent page would make sense.

**Comment by alexanderbez** [comment]
> No, search for messages sent by some address OR recevied to the same address. Think of an address as an account.

Yes, you can do this with events. To search for all transactions that contain _any_ message which has a sender A: `txs/message.sender={A}`. To search for all transactions where there is a `MsgSend` message which has a recipient B: `txs/transfer.recipient={B}`. In addition, you can combine these with the `AND` operand.

Each module has it's own `xx_events.md` file under the `spec/` directory. You can see all events there.

> Only in a very sloppy fashion. Say I finished on page 5 at the last request, now I need to re-fetch page 5 every time to see if something was added, wasting bandwidth. Or I would need to ask for page=&limit=1, and then make one more request to see if more than one was added. Again wasting bandwidth and creating unnecessary load on the server.

Sure. But in light of this "issue", it would be immediately clear to me that if this is your use case, then you're better off creating a subscription opposed to querying directly. Through a subscription, you'll get these txs in real time! That being said, do you have suggestions on alternatives? 

> Think "most recent N transactions". Even most recent page would make sense.

I see. I think we can amend the pagination logic to support such a query. In other words, support both pagination params and `count|num` or whatever we want to call it to get the last `n` txs.

**Comment by feri42** [comment]
> Yes, you can do this with events

This sounds great. Can I do something like: `message.sender={A} OR transfer.recipient={A}`? That is what I am looking for. Is there any documentation on the syntax of using the `AND` operand?

> you're better off creating a subscription 

I would be more than happy to have a subscription model available, but it does not seem like the Gaia-lite API supports subsriptions: https://cosmos.network/rpc/#/. I see that the Tendermint API supports subscription models, but there I have to worry about decoding (javascript).

> support both pagination params and count|num or whatever we want to call it to get the last n txs

Pagination is fine. What is missing is the starting point of pagination (e.g. block number or date).

**Comment by alexanderbez** [comment]
> This sounds great. Can I do something like: `message.sender={A} OR transfer.recipient={A}`? That is what I am looking for. Is there any documentation on the syntax of using the `AND` operand?

Unfortunately, I do not think Tendermint currently supports the `OR` operand. So replicate this, it would be two distinct queries or subscriptions.

When you query for txs through Gaia REST (`/txs?...) the `AND` is automatically done for you. e.g.
`/txs?message.sender=...&transfer.recipient=...&...`

> I would be more than happy to have a subscription model available, but it does not seem like the Gaia-lite API supports subsriptions: https://cosmos.network/rpc/#/. I see that the Tendermint API supports subscription models, but there I have to worry about decoding (javascript).

This is what I mean. Subscription through Tendermint RPC. Having it in gaia would serve no benefit really. There are javascript libraries out there that allow you to decode.

**Comment by feri42** [comment]
> Unfortunately, I do not think Tendermint currently supports the OR operand. So replicate this, it would be two distinct queries or subscriptions.

That is what I thought. But doing two queries is not too bad. It is much worse that there is no starting point for querying.

> There are javascript libraries out there that allow you to decode.

I am only aware of `js-amino`, which does not really work well. Do you know of any other?

**Comment by alexanderbez** [comment]
@jordansexton @marbar3778 do you know of other JS libraries that help with amino decoding?

**Comment by webmaster128** [comment]
> Pagination is fine. What is missing is the starting point of pagination (e.g. block number or date).

In raw Tendermint you can do both `tx.height=123` and `tx.height>123`. The first one is (undocumentedly) supported by the REST server (try `GET /txs?tx.height=123`). Unfortunately the second one results in a parse error like

> TxSearch: RPC error -32603 - Internal error: 
> parse error near digit (line 1 symbol 15 - line 1 symbol 16):
> "0"

It should not be super hard to pass `>` through the REST server just like `=`, given that Tendermint has support for both.
